Match List (static HTML)

trm313 opened the pull request:
First pass on Match List Items - currently static HTML written directly to template.

Review for: 

- Styling theme
- Data elements present to achieve requirements set in user story

Once these two items are agreed upon, I will create a data structure for returning match list data, and factor the components to receive data from appropriate object

trm313 commented:
The latest commit implements the dark theme styling to most of the app (although just need to touch up the Login / Signup page, specifically the text color of input fields). It also has some minor updates to the MatchListItem component

When reviewing:

- Any comments on the theming are more than welcome (see the master styles.css file for global variables and classes)
- Check out the data components being rendered in the match items, if we need to add/subtract, this is a good time to identify if we're meeting the criteria we need

Neoflash1979 commented:
@trm313 Wow, great work setting up the globals. I suck at css, I didn't even now about css variables, I had to look them up. Now I know. I like your color choices. The only one I'm not quite sure about is the purple accent color. I don't hate it, and can live with it, but I'm wondering if something else wouldn't look better.

As for the data showing in the match list, I think you did a good job of displaying gracefully the information that will be most useful for the users. I can't think of anything else that would be so important that it must be displayed on the card.

trm313 commented:
The purple was just to give it a little color splash (to help prevent all the data from blending together, and to draw the eye to the actions available) - not married to it at all. Any suggestions?

I finished refactoring the component so it dynamically builds itself from a data object. The mocked up data object is located in file match.data.ts -> I made an assumption that data for the MatchList would be returned from the server already customized to the logged in user (eg. sorted by relevance, etc.). 

Let me know what you guys think, and if we're good to go we can merge it in

Neoflash1979 commented on src/app/match-list/match-list.component.ts:
```diff
@@ -1,11 +1,15 @@
   styleUrls: ['./match-list.component.css']
 })
 export class MatchListComponent implements OnInit {
+  matches = matchData;
```
Although it is not needed, if you want to take advantage of Typescript's typing system, you could change line 12 to this: `matches: Match[] = matchData;`
This tells TypeScript that `matches` should be an array of `Match` objects.

Neoflash1979 commented on src/app/match-list/match.data.ts:
```diff
@@ -0,0 +1,142 @@
+export const  matchData = [
```
You could change line 1 to this: 
```
import { Match } from './match.model';

export const matchData: Match[] = [
```
This way, your editor will let you know if you are missing some properties, and TypeScript will throw an error if that is the case.

Neoflash1979 commented on src/app/match-list/match.model.ts:
```diff
@@ -0,0 +1,11 @@
+export class Match {
```
By making the above proposed change to match.data.ts, I quickly realized that the model was not updated for many of the properties you have included in your data. It should be updated accordingly. Here is my suggestion:
```
export class Match {
    constructor(
        public _id: string,
        public email: string,
        public username: string,
        public firstName: string,
        public lastName: string,
        public profileImageUrl: string,
        public devicePrefs: [{ name: string, image: string }],
        public bio: string,
        public lastActive: string,
        public timesFavorited: number,
        public numberTeams: number,
        public techPrefs: [{ language: string, proficiency: number }],
        public city?: string,
        public state?: string,
    ) { }
}
```
The `?` after the `city` and `state` properties tell TypeScript that they are optional since they are not being used on the data card.

Neoflash1979 commented on src/app/match-list/match-list-item/match-list-item.component.ts:
```diff
@@ -0,0 +1,30 @@
+    this.moreLangs = this.moreLanguages();
+  }
+
+  moreLanguages() {
```
Good job implementing this logic for the "more languages" tooltip. Just for fun, I did a bit of refactoring, feel free to implement my suggestion or not. Either way, nice work.
```
  moreLanguages(match: Match): string {
    if (match.techPrefs.length > 3) {
      return match.techPrefs.slice(3).map(tech => tech.language).join(', ');
    }
  }
```
Writing the function this way will make it easy to reuse if ever we need the same functionality elsewhere in the app. 
I'll add some more comments to other lines in this file to explain my proposed changes.
BTW, you should check out https://github.com/ryanmcdermott/clean-code-javascript it has opened my eyes.

Neoflash1979 commented on src/app/match-list/match-list-item/match-list-item.component.ts:
```diff
@@ -0,0 +1,30 @@
+  styleUrls: ['./match-list-item.component.css']
+})
+export class MatchListItemComponent implements OnInit {
+  @Input() match;
```
Consider changing line 9 to `@Input() match: Match;`, making use of typings.

Neoflash1979 commented on src/app/match-list/match-list-item/match-list-item.component.ts:
```diff
@@ -0,0 +1,30 @@
+export class MatchListItemComponent implements OnInit {
+  @Input() match;
+
+  moreLangs = '';
```
This is absolutely fine... a more TypeScripty way would be `moreLangs: string`.

Neoflash1979 commented on src/app/match-list/match-list-item/match-list-item.component.ts:
```diff
@@ -0,0 +1,30 @@
+
+  ngOnInit() {
+    console.log(this.match);
+    this.moreLangs = this.moreLanguages();
```
Since I refactored the `moreLanguages` function, it now needs to be called with the `this.match` parameter.
`this.moreLangs = this.moreLanguages(this.match);`

trm313 commented:
Awesome feedback Philippe! Thanks for looking at it so thoroughly.

I implemented all of your refactoring suggestions, as well as the cursor. I added a routerLink to the person's name navigating to address /users/<id> 

The only thing I didn't add yet is the country - played around with a few things but didn't like the design. I think you're right that perhaps timezone is the most effective. I'll play around with it a bit more. For now though, we might as well merge this so that everyone has access to the same CSS styling and everything
